### PR TITLE
Experimental support for fullscreen detection under KDE Plasma (Wayland)

### PR DESCRIPTION
**Pull request by LS-FCEFyN**
modified:   CMakeLists.txt
modified:   README.md
new file:     src/WallpaperEngine/Render/Drivers/Detectors/KDEWaylandFullScreenDetector.cpp
new file:     src/WallpaperEngine/Render/Drivers/Detectors/KDEWaylandFullScreenDetector.hpp
modified:   src/WallpaperEngine/Render/Drivers/Detectors/WaylandFullScreenDetector.cpp

This patch set introduces fullscreen detection on KDE Plasma via a KWin script that forwards internal KWin signals (e.g., `maximizedAboutToChange`, `maximizedChanged`, `minimizedChanged`, and geometry updates) to `Linux Wallpaper Engine` over D-Bus.

Although labeled as experimental, I have not encountered any functional issues so far. The “experimental” designation mainly reflects uncertainty around modern C++ idioms and conventions (for example, the use of anonymous namespaces versus `static` was something I was not aware of). As I refine my understanding of current conventions, further changes may follow.

### CODE REVIEW

File: src/WallpaperEngine/Render/Drivers/Detectors/KDEWaylandFullScreenDetector.cpp [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+        .horizontal = horizontal,
+        .vertical = vertical,
+        .fully = fully,
+        .appId = appId,
+    }
+    );
+
+    if (!windowKey.empty ()) {
+    m_activeWindowKey = windowKey;
+    }
```
[the remark this answers is not in the document]
Comment (LS-FCEFyN) [reply src/WallpaperEngine/Render/Drivers/Detectors/KDEWaylandFullScreenDetector.cpp:165]: KWin defines an "active window" as a the window to which user input is directed to (i.e. focused).
The KWin Maximize Detector script fires a signal when a window is activated, thus updating the `m_activeWindowKey` on in this C++ code is equivalent to tracking the currently  focused window.

[the remark this answers is not in the document]
Comment (LS-FCEFyN) [reply src/WallpaperEngine/Render/Drivers/Detectors/KDEWaylandFullScreenDetector.cpp:165]: I hadn't though about those edge cases quite profoundly, I'll re read the documentation run some tests and get back to you.
Any how if things do end up being the way you say they are the fix could most probably be made on the KWin script side.

[the remark this answers is not in the document]
Comment (LS-FCEFyN) [reply src/WallpaperEngine/Render/Drivers/Detectors/KDEWaylandFullScreenDetector.cpp:165]: You were right, even if a window is not focused signals are sent; however, after extensive testing with a C++ application that creates a window that changes its geometry between various windowed resolutions and also maximizes and unmaximizes itself I can confirm that if the focus is not stolen while the transformations are happening then nothing affects "--fullscreen-pause-only-active", thus the behavior is the expected behavior.

File: src/WallpaperEngine/Render/Drivers/Detectors/KDEWaylandFullScreenDetector.cpp [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+}
+
+void KDEWaylandFullScreenDetector::dispatchLoop () {
+    while (!m_stop.load (std::memory_order_acquire)) {
+	if (m_connection == nullptr) {
+	    break;
+	}
+
+	dbus_connection_read_write_dispatch (m_connection, 250);
+    }
```
[the remark this answers is not in the document]
Comment (LS-FCEFyN) [reply src/WallpaperEngine/Render/Drivers/Detectors/KDEWaylandFullScreenDetector.cpp]: I may be too tired because it is too late but before executing ` dbus_connection_read_write_dispatch()` I'm already checking if the member `m_connection` is a `nullptr`

[the remark this answers is not in the document]
Comment (LS-FCEFyN) [reply src/WallpaperEngine/Render/Drivers/Detectors/KDEWaylandFullScreenDetector.cpp]: Hmmmm that isnt quite right because member variable m_connection is not guarded by a mutex... that would add additional complexity that's not needed, the only time we modify `m_connection`  from outside the thread is when we join with the main thread on member function \ method `stopDBus()` which in turn is only called when the object is destructed...
A more appropriate fix would be to use the `reset()` method to reset the state to its original state, unregister the object path, unreference the connection and set  `m_connection = nullptr`andthen break the loop if the method `dbus_connection_read_write_dispatch(...)` returns false.

All that can be done safely because only the dispatch thread  operates on the connection once that thread has been created additionally  `stopDBus()` already has checks in place to ensure that we do not unregister and unreference things from a null connection.

We would technically end up with a zombie thread which isnt wrong rather technically untidy.... an alternative would be to switch to `std::async` but for the time being I'm only confident with concurrency and parallelism in Java.

[the remark this answers is not in the document]
Comment (LS-FCEFyN) [reply src/WallpaperEngine/Render/Drivers/Detectors/KDEWaylandFullScreenDetector.cpp]: In the strictest of senses you are right because it's not like we are holding onto resources, but we do still have an entry in the system table. It is technically dead/"defunct"but its parent cannot acknowledged its death to "reap" it because the only time we call `stopDBus()` is when the object is destroyed.

[the remark this answers is not in the document]
Comment (LS-FCEFyN) [reply src/WallpaperEngine/Render/Drivers/Detectors/KDEWaylandFullScreenDetector.cpp]: You learn something new everyday, I'll tidy up those mistakes you found in accordance to your suggestion and my proposed fix.

**Comment by LS-FCEFyN** [comment]
@coderabbitai review

**Comment by LS-FCEFyN** [comment]
@coderabbitai review

**Comment by LS-FCEFyN** [comment]
@coderabbitai review

it should be all good now, regarding your nitpick comment

`mutable on all D-Bus state members makes anythingFullscreen() const a side-effectful, non-idempotent operation.

Marking m_connection, m_windowStates, and m_activeWindowKey all mutable to satisfy the inherited const signature means calling anythingFullscreen() dispatches pending D-Bus messages and updates shared window state as a hidden side effect. Callers and the compiler alike will assume const implies no observable state change.

The root cause is the base-class signature (virtual bool anythingFullscreen() const). The cleanest long-term fix is to drop const there and propagate the change to all detectors. A cheaper alternative is to collect the mutable members into a single struct PollState and mark only that struct mutable, making the intent more explicit.`

I understand what you are trying to say, but technically the regular wayland detector uses pointer indirection to modify a member and keep track of things that are fullscreen, so while technically const is being honored in a way anythingFullScreen isnt really without side effects, that's why I declared my variables as `mutable` it allows me to honor const correctness while also being clearer rather than using indirections.
